### ISSUE TITLE
running packimage prevents nodes from booting.

### ISSUE THREAD

**Issue by ralphbellofatto** [xcat2/xcat-core#1940]
If we run the following command, which builds a 10GB diskless boot image and there was a previous valid image in that place, we are unable to boot the nodes that point to this image until the packimage completes.
```
packimage rh72_op5_compute
```

the node eventually boots up but reports this:
```
[Wed Oct  5 08:27:55 2016][   14.747784] dracut-initqueue[1470]: http://10.1.0.1//install/netboot/rhels7.2/ppc64le/rh72_op5_compute/rootimg.cpio.gz:                                                    
[Wed Oct  5 08:27:55 2016][   14.748376] dracut-initqueue[1470]: 2016-10-05 12:27:49 ERROR 403: Forbidden.                                                                                              
[Wed Oct  5 08:28:22 2016][   41.753150] dracut-initqueue[1470]: http://10.1.0.1//install/netboot/rhels7.2/ppc64le/rh72_op5_compute/rootimg.cpio.gz:                                                    
[Wed Oct  5 08:28:22 2016][   41.753823] dracut-initqueue[1470]: 2016-10-05 12:28:16 ERROR 403: Forbidden.                                                                                              
[Wed Oct  5 08:28:49 2016][   68.758139] dracut-initqueue[1470]: http://10.1.0.1//install/netboot/rhels7.2/ppc64le/rh72_op5_compute/rootimg.cpio.gz:                                                    
[Wed Oct  5 08:28:49 2016][   68.758699] dracut-initqueue[1470]: 2016-10-05 12:28:43 ERROR 403: Forbidden.                                                                                              
```

And there is nothing in the xcat node status to indicate that this is the hold up.

Since package takes about 10 minutes, this starts to cause operational problems when nodes are being rebooted.

I'm not sure what the expected behavior should be here.  There are a couple of possibilites.

- have pack image put the put the new compressed files into a temporary directory and when done, remove the original files and move the new files in its place.  This would have the advantage of cutting down the time that the booting node is waiting for the packimage to complete.
- detect the condition and change the lsdef status and the nodestat results during the boot sequence to indicate a delay is in progress and why.

- when performing the pack image, warn the operator that what they are doing will prevent nodes from booting if they are currently provisioned with that diskless image.  This might encourage using an a/b approach to image directories and properly direct the administrator to avoid using one that is currently in play.

Right now, the system generates a bit of confusion, this is more more because of the time it takes to pack a large image and causes us to have higher admin costs due to the confusion.

-

**Comment by whowutwut** [comment]
@ralphbellofatto  Yes, the error makes sense because it's the packed image that is going to be served to the diskless compute nodes and if you are writing it, it's not complete.  

However, I do think that this something that happens in the development/test environment but I don't think this is something that would occur on a production system, because a osimage that is being used in production, should not be overwritten.  If the admin is creating a new one, it would be done as a different osimage or on another system to test/verify before moving into production. 

Although that being said,  I do like the first option that you recommended as that would reduce the chance of this happening, which is also good.

**Comment by ralphbellofatto** [comment]
ORNL e-mailed me on this issue and suggested pigz be installed on the system.

I did that and the compression time is now: 

real    1m29.727s

They also mentioned their method of dealing with this, It involves keeping the same osimage name but putting a date/time stamp on the rootimage directory, which handles the issue for them.

We will probably adapt this in the future for just simple updates, and a separate osimage name for more complicated ones.

**Comment by immarvin** [comment]
hi @whowutwut, for the first option, there might still be confusing caused by the race condition of the timing of the following steps: 
1. issue a provision with osimage A with compressed rootimg "A-root1"
2. "packimage A"  generate a new compressed rootimg "A-root2" 
it is hard for user to figure out which rootimg tarball  is actually provisioned on the node.

another option is to copy the compressed rootimg to "/tftpboot/xcat/osimage/xxxx" on "nodeset", since "nodeset" is the command to specify the next provision osimage.

**Comment by ralphbellofatto** [comment]
The tarball that will provision the node the next time it boots is in the initrd attribute of the lsdef for that node.

What I think is missing for the solution of adding a date to the rootimg dir is a way to tell what image was used for an already booted image.  

Packimage can take long time, so if it is done on node set, it should be idempotememt and only perform the packimage step if genimage was run again, or better yet, the rootimgdir was altered in any way.

**Comment by whowutwut** [comment]
Actually, @cxhong  made a change to include a unique identifier in the `/opt/xcat/xcatinfo` file : https://github.com/xcat2/xcat-core/pull/1297   There is an `xcatprobe image -c` option that you can use to check for duplicate osimage deployed ...  would like feedback for that probe command.

**Comment by ralphbellofatto** [comment]
duplicate, or being able to match up an image deployed to an osimage record and the specific osimage directory??

**Comment by gurevichmark** [comment]
It checks to make sure the diskless compute node is running with the same image as defined in osimage definition for that node.

**Comment by ralphbellofatto** [comment]
how does it perform this check?

**Comment by gurevichmark** [comment]
It compared the value of IMAGEUUID in /opt/xcat/xcatinfo file on the running compute node with the value of IMAGEUUID in {ROOTIMGDIR}/rootimg/opt/xcat/xcatinfo file, for that node, on the MN.

**Comment by ralphbellofatto** [comment]
ok, that tells us that we are matched with what is currently in the rootimage directory that the system is pointed to.

If we do the change of the rootimage dir everytime we generate a changed image, is there a way to tell in the installed image what rootimage that it is matched up with or do we need to search all our root image directories for a matching uuid...

 I.e. is the full path to the rootimage directory that this image was booted with stored anywhere on the running image and can it be easily be displayed?

**Comment by ralphbellofatto** [comment]
interesting results from xcat probe:

It blathers a bunch warnings about the switches:
c460c090 has no provision method defined. No image consistency verification will be performed.[WARN]
c460csw01 has no provision method defined. No image consistency verification will be perfor...[WARN]
c460f01ib1 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460f02ib1 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460f03ib1 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460f04ib1 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460f06ib1 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460login01 is not diskless. No image consistency verification will be performed.             [WARN]
c460login02 has no provision method defined. No image consistency verification will be perf...[WARN]
c460nfs01 has no provision method defined. No image consistency verification will be perfor...[WARN]
c460nfs02 has no provision method defined. No image consistency verification will be perfor...[WARN]
c460tors01 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460tors02 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460tors03 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460tors04 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460tors05 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460tors06 has no provision method defined. No image consistency verification will be perfo...[WARN]
c460ufm01 has no provision method defined. No image consistency verification will be perfor...[WARN]

I'm curious as to why it does this since this seems to be configured as a switch and not any kind of compute node:
```

[root@c460mgt01 otherpkgs]# lsdef c460tors01
Object name: c460tors01
    groups=switch,all,ecswitchs
    hostnames=c460f01u42
    ip=10.1.1.20
    mac=8c:ea:1b:12:d4:c0
    mgt=switch
    nodetype=switch
    password=CumulusLinux!
    postbootscripts=syncfiles
    postscripts=syslog,remoteshell,syncfiles,ospkgs,otherpkgs
    protocol=ssh
    rack=c460f01
    serial=EC1631000223
    switch=c460csw01
    switchport=29
    switchtype=cumulus
    unit=u42
    usercomment=expermental_cumulus_switch
    username=cumulus
```

Is there something we reconfigured here?  Or are these warnings that we have to train our admin staff to ignore?

**Comment by gurevichmark** [comment]
Just my lack of experience. :-)
I will add code to ignore switches. You can also use a `-n` flag to specify noderange.

**Comment by gurevichmark** [comment]
Fixed by pull request #1956

**Comment by whowutwut** [comment]
Closing issue
